### PR TITLE
Frontend testing documentation

### PR DESCRIPTION
**Pull request by nmashton**

### CODE REVIEW

**Comment by dpoirier** [comment]
These pulls are easier to review if the built HTML/CSS etc files are not included in the pull.  We typically run a build and commit the changes to develop after a pull has been merged.

File: source/frontend-testing.rst [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+
+It also reflects a series of choices that were basically arbitrary because of
+the lack of anything like a standard or clear best choice (e.g. the
+decision to use one test runner or assertion library over alternatives).
+
+These choices are still being experimented with, and developers should feel
+free to suggest revisions if they're found to perform poorly in practice.
+
+
+Testing libraries: overview
```
Comment (ironfroggy) [inline source/frontend-testing.rst]: Title Case Please

File: source/frontend-testing.rst [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+-----
+
+`Mocha <https://mochajs.org/#features>`_ is a simple and modular test library.
+Unlike `Jasmine <http://jasmine.github.io/>`_, Mocha doesn't come with an
+assertion library, facilities for mocks and spies, and so on. Instead it provides
+a framework for writing and executing tests and lets you choose other libraries
+to meet those needs.
+
+To create a Mocha test that will be recognized by our Gulp build process,
+create a file with the extension ``.spec.js`` and place it in a ``test/`` subdir
```
Comment (ironfroggy) [inline source/frontend-testing.rst]: Would it be unreasonable to change this pattern from `*.spec.js` to `test_*.js` just so it is more consistent with the existing backend tests?

File: source/frontend-testing.rst [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+    {
+      "presets": ["es2015"],
+      "plugins": ["transform-react-jsx"]
+    }
+
+This allows us to not only write and test ES2015 and JSX applications but to write
+our tests themselves in ES2015 and JSX.
+
+React tests with jsdom
+----------------------
```
Comment (ironfroggy) [inline source/frontend-testing.rst]: Let's include in this section that the principles can apply to other frontend libraries, not just React.

File: source/frontend-testing.rst [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+To check for changes like that, use an async test and inject the ``done`` callback
+into the appropriate React component lifecycle method::
+
+    it('updates when the store changes', (done) => {
+      let old_componentDidUpdate = YourComponent.prototype.componentDidUpdate;
+      YourComponent.prototype.componentDidUpdate = () => {
+        // this lifecycle method will be called when the update has happened
+        YourComponent.prototype.componentDidUpdate = old_componentDidUpdate;
+        done();
+      };
```
Comment (ironfroggy) [inline source/frontend-testing.rst]: Can we wrap this pattern into a utility that does it for us?

**Comment by rlconley** [comment]
:dromedary_camel:

**Pull request merged**